TLS Bootstrapping

surajssd opened the pull request:
This PR enables TLS bootstrap for Lokomotive.

**NOTE**: Look at the PR per commit to look at the changes.

Fixes https://github.com/kinvolk/lokomotive/issues/177

---

- In kube-apiserver config:

  - Add `Node` authoriser.
  - Enable `NodeRestriction` admission plugin.
  - Enable bootstrap token auth.

- In kube-controller-manager:

  - Enable tokencleaner controller:
  - Set cluster signing duration flag.

- Remove the ClusterRoleBinding `system-nodes` which assigned group `system:nodes` permission to role `system:node`.

- This PR removes all the certs generated statically for kubelet. These certs are no longer needed with bootstrap kubelet.

- Add terraform plugin `random` to bootkube which will generate the random strings for bootstrap token.

- Add bootstrap-secrets chart

  - Add chart that will install bootstrap secrets in `kube-system` namespace corresponding to a node.

  - Add variable `bootstrap_tokens` to bootkube so that platforms can pass a list of `token-id` and `token-secret` corresponding to each node in the cluster.

  - Add ClusterRoleBindings so as to enable kubelet bootstrapping mechanism.


- In kubelet:

  - Provide the bootrap kubeconfig via flag `--bootstrap-kubeconfig`.

  - Provide a path to store downloaded kubeconfig via existing `--kubeconfig` flag. This place has to writable.

  - Added flag to rotate ceritificates.

  - Change the path of kubeconfig to delete node.


- For packet: Generate bootstrap tokens per node

  - Create bootstrap kubeconfig, with random token id and secret, per node.

  - Add bootstrap kubeconfig template.

  - Output CA cert, API server endpoint and `bootstrap-secrets` chart values in controller module.

  - Pass CA cert and API server endpoint to worker pools, so that they can construct bootstrap kubeconfig. To enable this data flow add variables `ca_cert` and `apiserver`.

  - Pass tokens from worker pools to controller pool, so that it combined with controller pool tokens and can be passed to `bootstrap-secrets` chart in bootkube. To enable this data flow add variable `worker_bootstrap_tokens` in controller module and output `worker_bootstrap_tokens` in worker module.

  - Don't pass statically generated kubeconfig to the workers.


- cluster-apply: Apply bootstrap-secrets helm chart without node bootup check

  - Bootstrap secrets should be available so that nodes join the cluster. So this chart should be installed without node readiness checks.


- For aws: Generate bootstrap tokens per node

  - Create bootstrap kubeconfig, with random token id and secret, per node.

  - Add bootstrap kubeconfig template.

  - Output CA cert, API server endpoint and `bootstrap-secrets` chart values in controller module.

  - Pass CA cert and API server endpoint to worker pools, so that they can construct bootstrap kubeconfig. To enable this data flow add variables `ca_cert` and `apiserver`.

  - Pass tokens from worker pools to controller pool, so that it combined with controller pool tokens and can be passed to `bootstrap-secrets` chart in bootkube. To enable this data flow add variable `worker_bootstrap_tokens` in controller module and output `worker_bootstrap_tokens` in worker module.

  - Don't pass statically generated kubeconfig to the workers.

  - Pass kubeconfig to controller nodes via ignition config.

- For baremetal: Generate bootstrap tokens per node

  - Create bootstrap kubeconfig, with random token id and secret, per node.
  
  - Add bootstrap kubeconfig template.


---

### Replace a token

This helped in in regenerating the token and replacing the node:

```
export poolname=pool-1
export workernumber=2
export clustername=suraj-lk-cluster

cd assets/terraform
kubectl drain $clustername-$poolname-worker-$workernumber
kubectl delete node $clustername-$poolname-worker-$workernumber
terraform taint module.worker-$poolname.random_string.bootstrap-token-id[$workernumber]
terraform taint module.worker-$poolname.random_string.bootstrap-token-secret[$workernumber]
terraform taint module.worker-$poolname.packet_device.nodes[$workernumber]
```

surajssd commented on assets/lokomotive-kubernetes/bootkube/resources/bootstrap-manifests/bootstrap-controller-manager.yaml:
```diff
@@ -18,6 +18,8 @@ spec:
     - --cluster-signing-key-file=/etc/kubernetes/secrets/ca.key
     - --configure-cloud-routes=false
+    - --controllers=*,tokencleaner,bootstrapsigner
+    - --experimental-cluster-signing-duration=1h
```
What should be the ideal time to renew the certificate? Days/Hours? If hours how many?

invidian replied:
Why not use default?

invidian replied:
Also, the `--controllers` flag is not really needed I think. Can you elaborate why you add it?

surajssd replied:
Docs mention that you should add it. So here https://kubernetes.io/docs/reference/access-authn-authz/bootstrap-tokens/#enabling-bootstrap-token-authentication.

Also 

> --controllers stringSlice     Default: [*]
> 
>  | A list of controllers to enable. '*' enables all on-by-default controllers, 'foo' enables the controller named 'foo', '-foo' disables the controller named 'foo'.All controllers: attachdetach, bootstrapsigner, cloud-node-lifecycle, clusterrole-aggregation, cronjob, csrapproving, csrcleaner, csrsigning, daemonset, deployment, disruption, endpoint, endpointslice, garbagecollector, horizontalpodautoscaling, job, namespace, nodeipam, nodelifecycle, persistentvolume-binder, persistentvolume-expander, podgc, pv-protection, pvc-protection, replicaset, replicationcontroller, resourcequota, root-ca-cert-publisher, route, service, serviceaccount, serviceaccount-token, statefulset, tokencleaner, ttl, ttl-after-finished Disabled-by-default controllers: bootstrapsigner, tokencleaner

src: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

see the last line says that `Disabled-by-default controllers: bootstrapsigner, tokencleaner` are disabled by default and we need to enable them.

invidian replied:
Right. I'm not exactly sure how this configmap signing process works, but it seems it's something we should use :+1: thanks for clarifying.

surajssd replied:
It turns out we don't need `bootstrapsigner`, it is a kubeadm feature. Since in kubeadm the way for a worker machine to join the cluster is by using the token, there is no way for the worker to know if the apiserver it is trying to join is a valid one and thus this feature comes in handy.

Also the worker node has no CA data and this configmap helps in acquiring that as well. We don't have such problems though.

invidian commented:
>Right now we use same token for each node. Which is not a big deal of a problem if you think about it. The real problem is creating a new token and distributing it across nodes. How do we achieve it later, if the token is leaked accidentally?

What I would do is to create a token per worker group and then collect them in bootkube module and create them from there. 

Also, in case the token leaks, one can taint the random resources in Terraform and tokens will be recreated. And so should be worker groups then.

surajssd commented:
@invidian 
> Also, in case the token leaks, one can taint the random resources in Terraform and tokens will be recreated. And so should be worker groups then.

Can this be triggered from lokoctl? Or can we think about a feature of doing it like that for lokoctl?

I would have ideally wanted a per node token. And a mechanism to ship bootstrap token without node recreation.

invidian commented:
>Can this be triggered from lokoctl? Or can we think about a feature of doing it like that for lokoctl?

Yes, I can imagine we could implement it as a feature in `lokoctl`. To be able to manage the tokens somehow.

>I would have ideally wanted a per node token. 

I don't think that's possible. We use Ignition configuration **per worker group** to utilize cloud node groups capabilities. 

>And a mechanism to ship bootstrap token without node recreation.

I think it's OK, if we don't allow updating it on the nodes, as after the node registers, it should use obtained kubeconfig anyway, not the bootstrap token. I wonder for how long the certificates are issued. Because if it's like 90 days, then node would have to be down for this period to require re-using bootstrap token. I'm not even sure if that's implemented.

surajssd commented:
This helped in in regenerating the token and replacing the node:

```
export poolname=pool-1
export workernumber=2
export clustername=suraj-lk-cluster

cd assets/terraform
kubectl drain $clustername-$poolname-worker-$workernumber
kubectl delete node $clustername-$poolname-worker-$workernumber
terraform taint module.worker-$poolname.random_string.bootstrap-token-id[$workernumber]
terraform taint module.worker-$poolname.random_string.bootstrap-token-secret[$workernumber]
terraform taint module.worker-$poolname.packet_device.nodes[$workernumber]
```

surajssd commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kubernetes/templates/kube-controller-manager.yaml:
```diff
@@ -60,6 +60,8 @@ spec:
         - --cluster-signing-key-file=/etc/kubernetes/secrets/ca.key
         - --configure-cloud-routes=false
+        - --controllers=*,tokencleaner
+        - --experimental-cluster-signing-duration=1h
```
This is still an issue of discussion. What should be the duration by default here?

invidian replied:
Do we have any strong reason **not to use** upstream default?

surajssd replied:
Upstream default is 1yr which is equal to the default expiry duration of the root CA.

invidian replied:
IMO it's fine like this, as we still have a benefit of using TLS bootstrapping, as it allows to enable node restrictions and each node use unique certificate.

surajssd commented:
I deployed a cluster with current master 461abbf91f1ae4986bcc5acc6ea8b63bb943fbe9 and then tried to upgrade it with current PR and I see following changes:

<details><summary>Details</summary>
<p>

```hcl
$ lokoctl cluster apply -v
...
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create
  ~ update in-place
  - destroy
 <= read (data resources)

Terraform will perform the following actions:

  # module.packet-my-cluster.data.ct_config.controller-ignitions[0] will be read during apply
  # (config refers to values not yet known)
 <= data "ct_config" "controller-ignitions"  {
      + content  = (known after apply)
      + id       = (known after apply)
      + platform = "packet"
      + rendered = (known after apply)
      + snippets = []
    }

  # module.packet-my-cluster.data.ct_config.controller-install-ignitions[0] will be read during apply
  # (config refers to values not yet known)
 <= data "ct_config" "controller-install-ignitions"  {
      + content  = (known after apply)
      + id       = (known after apply)
      + rendered = (known after apply)
    }

  # module.packet-my-cluster.data.template_file.bootstrap-kubeconfig[0] will be read during apply
  # (config refers to values not yet known)
 <= data "template_file" "bootstrap-kubeconfig"  {
      + id       = (known after apply)
      + rendered = (known after apply)
      + template = <<~EOT
            apiVersion: v1
            kind: Config
            clusters:
            - name: local
              cluster:
                server: ${server}
                certificate-authority-data: ${ca_cert}
            users:
            - name: kubelet
              user:
                token: ${token_id}.${token_secret}
            contexts:
            - context:
                cluster: local
                user: kubelet
        EOT
      + vars     = {
          + "ca_cert"      = "[REDACTED]"
          + "server"       = "https://my-cluster-private.example.com:6443"
          + "token_id"     = (known after apply)
          + "token_secret" = (known after apply)
        }
    }

  # module.packet-my-cluster.packet_device.controllers[0] will be updated in-place
  ~ resource "packet_device" "controllers" {
        access_private_ipv4              = "10.88.81.1"
        access_public_ipv4               = "1.2.3.4"
        access_public_ipv6               = "2604:1380:1000:cd00::1"
        always_pxe                       = false
        billing_cycle                    = "hourly"
        created                          = "2020-06-19T10:13:11Z"
        deployed_facility                = "sjc1"
        facilities                       = [
            "sjc1",
        ]
        force_detach_volumes             = false
        hostname                         = "my-cluster-controller-0"
        id                               = "02369f5a-f2fb-4630-8278-2c58b2e52182"
        locked                           = false
        network                          = [
            {
                address = "1.2.3.4"
                cidr    = 31
                family  = 4
                gateway = "1.2.3.5"
                public  = true
            },
            {
                address = "2604:1380:1000:cd00::1"
                cidr    = 127
                family  = 6
                gateway = "2604:1380:1000:cd00::"
                public  = true
            },
            {
                address = "10.88.81.1"
                cidr    = 31
                family  = 4
                gateway = "10.88.81.0"
                public  = false
            },
        ]
        network_type                     = "layer3"
        operating_system                 = "flatcar_stable"
        plan                             = "t1.small.x86"
        ports                            = [
            {
                bonded = true
                id     = "f0f53195-8600-4a5f-9a23-eb713e75d1fb"
                mac    = ""
                name   = "bond0"
                type   = "NetworkBondPort"
            },
            {
                bonded = true
                id     = "78ed0f56-35cc-4f5c-8d76-e06f07946b14"
                mac    = "0c:c4:7a:b5:86:ea"
                name   = "eth0"
                type   = "NetworkPort"
            },
            {
                bonded = true
                id     = "2a75c5c7-f65f-43a3-93d8-7408f2e928e2"
                mac    = "0c:c4:7a:b5:86:eb"
                name   = "eth1"
                type   = "NetworkPort"
            },
        ]
        project_id                       = "436c24d4-770d-46c7-a610-8c32d7e8fe22"
        public_ipv4_subnet_size          = 31
        ssh_key_ids                      = [
            "[REDACTED]",
        ]
        state                            = "active"
      ~ tags                             = [
          - "lokoctl-version:v0.1.0-489-g461abbf9",
          + "lokoctl-version:v0.1.0-500-gd36430b7",
        ]
        updated                          = "2020-06-19T10:20:18Z"
      ~ user_data                        = (sensitive value)
        wait_for_reservation_deprovision = false
    }

  # module.packet-my-cluster.random_string.bootstrap-token-id[0] will be created
  + resource "random_string" "bootstrap-token-id" {
      + id          = (known after apply)
      + length      = 6
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

  # module.packet-my-cluster.random_string.bootstrap-token-secret[0] will be created
  + resource "random_string" "bootstrap-token-secret" {
      + id          = (known after apply)
      + length      = 16
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

  # module.packet-my-cluster.template_dir.calico_host_protection will be created
  + resource "template_dir" "calico_host_protection" {
      + destination_dir = "../cluster-assets/charts/kube-system/calico-host-protection"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/packet/flatcar-linux/kubernetes/calico-host-protection"
    }

  # module.worker-pool-1.data.ct_config.ignitions[0] will be read during apply
  # (config refers to values not yet known)
 <= data "ct_config" "ignitions"  {
      + content  = (known after apply)
      + id       = (known after apply)
      + platform = "packet"
      + rendered = (known after apply)
      + snippets = []
    }

  # module.worker-pool-1.data.ct_config.install-ignitions[0] will be read during apply
  # (config refers to values not yet known)
 <= data "ct_config" "install-ignitions"  {
      + content  = (known after apply)
      + id       = (known after apply)
      + rendered = (known after apply)
    }

  # module.worker-pool-1.data.template_file.bootstrap-kubeconfig[0] will be read during apply
  # (config refers to values not yet known)
 <= data "template_file" "bootstrap-kubeconfig"  {
      + id       = (known after apply)
      + rendered = (known after apply)
      + template = <<~EOT
            apiVersion: v1
            kind: Config
            clusters:
            - name: local
              cluster:
                server: ${server}
                certificate-authority-data: ${ca_cert}
            users:
            - name: kubelet
              user:
                token: ${token_id}.${token_secret}
            contexts:
            - context:
                cluster: local
                user: kubelet
        EOT
      + vars     = {
          + "ca_cert"      = "[REDACTED]"
          + "server"       = "https://my-cluster-private.example.com:6443"
          + "token_id"     = (known after apply)
          + "token_secret" = (known after apply)
        }
    }

  # module.worker-pool-1.packet_device.nodes[0] will be updated in-place
  ~ resource "packet_device" "nodes" {
        access_private_ipv4              = "10.88.81.3"
        access_public_ipv4               = "2.3.4.5"
        access_public_ipv6               = "2604:1380:1000:cd00::3"
        always_pxe                       = false
        billing_cycle                    = "hourly"
        created                          = "2020-06-19T10:13:10Z"
        deployed_facility                = "sjc1"
        facilities                       = [
            "sjc1",
        ]
        force_detach_volumes             = false
        hostname                         = "my-cluster-pool-1-worker-0"
        id                               = "0abff247-04a3-4e25-ad96-ad0d0b7c28f7"
        locked                           = false
        network                          = [
            {
                address = "2.3.4.5"
                cidr    = 31
                family  = 4
                gateway = "2.3.4.6"
                public  = true
            },
            {
                address = "2604:1380:1000:cd00::3"
                cidr    = 127
                family  = 6
                gateway = "2604:1380:1000:cd00::2"
                public  = true
            },
            {
                address = "10.88.81.3"
                cidr    = 31
                family  = 4
                gateway = "10.88.81.2"
                public  = false
            },
        ]
        network_type                     = "layer3"
        operating_system                 = "flatcar_stable"
        plan                             = "c2.medium.x86"
        ports                            = [
            {
                bonded = true
                id     = "bc3a866b-c6d9-4477-987d-5d5c2ed20885"
                mac    = ""
                name   = "bond0"
                type   = "NetworkBondPort"
            },
            {
                bonded = true
                id     = "8e5e8c91-456d-4b5a-b08d-f43e321a3b23"
                mac    = "50:6b:4b:b7:16:b6"
                name   = "eth0"
                type   = "NetworkPort"
            },
            {
                bonded = true
                id     = "89d209db-6f1b-4501-bb87-de4a787dd864"
                mac    = "50:6b:4b:b7:16:b7"
                name   = "eth1"
                type   = "NetworkPort"
            },
        ]
        project_id                       = "436c24d4-770d-46c7-a610-8c32d7e8fe22"
        public_ipv4_subnet_size          = 31
        ssh_key_ids                      = [
            "[REDACTED]",
        ]
        state                            = "active"
      ~ tags                             = [
          - "lokoctl-version:v0.1.0-489-g461abbf9",
          + "lokoctl-version:v0.1.0-500-gd36430b7",
        ]
        updated                          = "2020-06-19T10:23:04Z"
      ~ user_data                        = (sensitive value)
        wait_for_reservation_deprovision = false
    }

  # module.worker-pool-1.random_string.bootstrap-token-id[0] will be created
  + resource "random_string" "bootstrap-token-id" {
      + id          = (known after apply)
      + length      = 6
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

  # module.worker-pool-1.random_string.bootstrap-token-secret[0] will be created
  + resource "random_string" "bootstrap-token-secret" {
      + id          = (known after apply)
      + length      = 16
      + lower       = true
      + min_lower   = 0
      + min_numeric = 0
      + min_special = 0
      + min_upper   = 0
      + number      = true
      + result      = (known after apply)
      + special     = false
      + upper       = false
    }

  # module.packet-my-cluster.module.bootkube.local_file.bootstrap-secrets will be created
  + resource "local_file" "bootstrap-secrets" {
      + content              = (known after apply)
      + directory_permission = "0777"
      + file_permission      = "0777"
      + filename             = "../cluster-assets/charts/kube-system/bootstrap-secrets.yaml"
      + id                   = (known after apply)
    }

  # module.packet-my-cluster.module.bootkube.local_file.kubeconfig-kubelet will be destroyed
  - resource "local_file" "kubeconfig-kubelet" {
      - content              = <<~EOT
            apiVersion: v1
            kind: Config
            clusters:
            - name: local
              cluster:
                server: https://my-cluster-private.example.com:6443
                certificate-authority-data: [REDACTED]
            users:
            - name: kubelet
              user:
                client-certificate-data: [REDACTED]
                client-key-data: [REDACTED]
            contexts:
            - context:
                cluster: local
                user: kubelet
        EOT -> null
      - directory_permission = "0777" -> null
      - file_permission      = "0777" -> null
      - filename             = "../cluster-assets/auth/kubeconfig-kubelet" -> null
      - id                   = "87b108749e4613132f302090c2250614e582e0a1" -> null
    }

  # module.packet-my-cluster.module.bootkube.local_file.kubelet-crt will be destroyed
  - resource "local_file" "kubelet-crt" {
      - content              = <<~EOT
            -----BEGIN CERTIFICATE-----
            [REDACTED]
            -----END CERTIFICATE-----
        EOT -> null
      - directory_permission = "0777" -> null
      - file_permission      = "0777" -> null
      - filename             = "../cluster-assets/tls/kubelet.crt" -> null
      - id                   = "b3ca8228871443057e4795d62a671cb2325593bb" -> null
    }

  # module.packet-my-cluster.module.bootkube.local_file.kubelet-key will be destroyed
  - resource "local_file" "kubelet-key" {
      - content              = <<~EOT
            -----BEGIN RSA PRIVATE KEY-----
            [REDACTED]
            -----END RSA PRIVATE KEY-----
        EOT -> null
      - directory_permission = "0777" -> null
      - file_permission      = "0777" -> null
      - filename             = "../cluster-assets/tls/kubelet.key" -> null
      - id                   = "43c482d7068e1b5d1776874e540a3a5b209e3fac" -> null
    }

  # module.packet-my-cluster.module.bootkube.template_dir.bootstrap-manifests will be created
  + resource "template_dir" "bootstrap-manifests" {
      + destination_dir = "../cluster-assets/bootstrap-manifests"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/bootstrap-manifests"
      + vars            = {
          + "cloud_provider"                = ""
          + "etcd_servers"                  = "https://my-cluster-etcd0.dev.lokomotive-k8s.net:2379"
          + "kube_apiserver_image"          = "k8s.gcr.io/kube-apiserver:v1.18.3"
          + "kube_controller_manager_image" = "k8s.gcr.io/kube-controller-manager:v1.18.3"
          + "kube_scheduler_image"          = "k8s.gcr.io/kube-scheduler:v1.18.3"
          + "pod_cidr"                      = "10.2.0.0/16"
          + "service_cidr"                  = "10.3.0.0/16"
          + "trusted_certs_dir"             = "/usr/share/ca-certificates"
        }
    }

  # module.packet-my-cluster.module.bootkube.template_dir.bootstrap-secrets will be created
  + resource "template_dir" "bootstrap-secrets" {
      + destination_dir = "../cluster-assets/charts/kube-system/bootstrap-secrets"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/charts/bootstrap-secrets"
    }

  # module.packet-my-cluster.module.bootkube.template_dir.calico will be created
  + resource "template_dir" "calico" {
      + destination_dir = "../cluster-assets/charts/kube-system/calico"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/charts/calico"
    }

  # module.packet-my-cluster.module.bootkube.template_dir.kube-apiserver will be created
  + resource "template_dir" "kube-apiserver" {
      + destination_dir = "../cluster-assets/charts/kube-system/kube-apiserver"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/charts/kube-apiserver"
    }

  # module.packet-my-cluster.module.bootkube.template_dir.kubernetes will be created
  + resource "template_dir" "kubernetes" {
      + destination_dir = "../cluster-assets/charts/kube-system/kubernetes"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/charts/kubernetes"
    }

  # module.packet-my-cluster.module.bootkube.template_dir.pod-checkpointer will be created
  + resource "template_dir" "pod-checkpointer" {
      + destination_dir = "../cluster-assets/charts/kube-system/pod-checkpointer"
      + id              = (known after apply)
      + source_dir      = "../lokomotive-kubernetes/bootkube/resources/charts/pod-checkpointer"
    }

  # module.packet-my-cluster.module.bootkube.tls_cert_request.kubelet will be destroyed
  - resource "tls_cert_request" "kubelet" {
      - cert_request_pem = <<~EOT
            -----BEGIN CERTIFICATE REQUEST-----
            [REDACTED]
            -----END CERTIFICATE REQUEST-----
        EOT -> null
      - id               = "bc1e3d773d9faf14264da127ac8e723cad82e606" -> null
      - key_algorithm    = "RSA" -> null
      - private_key_pem  = (sensitive value)

      - subject {
          - common_name    = "kubelet" -> null
          - organization   = "system:nodes" -> null
          - street_address = [] -> null
        }
    }

  # module.packet-my-cluster.module.bootkube.tls_locally_signed_cert.kubelet will be destroyed
  - resource "tls_locally_signed_cert" "kubelet" {
      - allowed_uses          = [
          - "key_encipherment",
          - "digital_signature",
          - "server_auth",
          - "client_auth",
        ] -> null
      - ca_cert_pem           = "05bb582154ce7258551c8eba1bfc0a71d2fd10d1" -> null
      - ca_key_algorithm      = "RSA" -> null
      - ca_private_key_pem    = (sensitive value)
      - cert_pem              = <<~EOT
            -----BEGIN CERTIFICATE-----
            [REDACTED]
            -----END CERTIFICATE-----
        EOT -> null
      - cert_request_pem      = "b6b17a5743ed0549bd29178536a8ee853d43574d" -> null
      - early_renewal_hours   = 0 -> null
      - id                    = "189863159367432705978437838633353101998" -> null
      - ready_for_renewal     = false -> null
      - validity_end_time     = "2021-06-19T15:43:07.92944791+05:30" -> null
      - validity_period_hours = 8760 -> null
      - validity_start_time   = "2020-06-19T15:43:07.92944791+05:30" -> null
    }

  # module.packet-my-cluster.module.bootkube.tls_private_key.kubelet will be destroyed
  - resource "tls_private_key" "kubelet" {
      - algorithm                  = "RSA" -> null
      - ecdsa_curve                = "P224" -> null
      - id                         = "ce5fc96d008d61bc2c0db0c72bc9665e86b28ba9" -> null
      - private_key_pem            = (sensitive value)
      - public_key_fingerprint_md5 = "82:c6:8f:ec:b6:39:df:22:7d:dc:5a:02:52:be:e8:0d" -> null
      - public_key_openssh         = <<~EOT
            [REDACTED]
        EOT -> null
      - public_key_pem             = <<~EOT
            -----BEGIN PUBLIC KEY-----
            [REDACTED]
            -----END PUBLIC KEY-----
        EOT -> null
      - rsa_bits                   = 2048 -> null
    }

Plan: 12 to add, 2 to change, 6 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

Do you want to proceed with cluster apply? [type "yes" to continue]: hellno
INFO[0085] Cluster apply cancelled                       args="[]" command="lokoctl cluster apply"
```




</p>
</details>

surajssd commented:
The PR is open @iaguis @ipochi @invidian

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+locals {
+  controller_bootstrap_tokens = [
+    for index in range(var.controller_count) : {
```
I'd use single token for all controllers, to support #28. And also to not make controllers much different from worker pools.

surajssd replied:
The tokens are not assigned to worker pools, but to worker. Token per node.

invidian replied:
I'm saying we should use token per worker pool, so all worker pools use the same configuration.

surajssd replied:
But that causes the entire pool to be recreated thus losing data. How do we mitigate that?

surajssd replied:
> Note, that we will replace nodes, as part of upgrade process, as we discussed some time ago.

That was only for ephemeral nodes. What about the nodes with stateful apps?

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+# Generate a cryptographically random token id (public)
```
```suggestion
# Generate a cryptographically random token id (public).
```

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+  special = false
+}
+
+# Generate a cryptographically random token secret
```
```suggestion
# Generate a cryptographically random token secret.
```

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+}
+
+
+data "template_file" "bootstrap-kubeconfig" {
```
I suggest we use `templatefile` instead, as per #196.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/bootstrap-manifests/bootstrap-apiserver.yaml:
```diff
@@ -15,11 +15,12 @@ spec:
     - --allow-privileged=true
     - --anonymous-auth=false
-    - --authorization-mode=RBAC
+    - --authorization-mode=Node,RBAC
```
In commit message, it should be `authorizer`, not `authoriser`, as in https://kubernetes.io/docs/reference/access-authn-authz/node/.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/bootstrap-manifests/bootstrap-apiserver.yaml:
```diff
@@ -15,11 +15,12 @@ spec:
     - --client-ca-file=/etc/kubernetes/secrets/ca.crt
     - --cloud-provider=${cloud_provider}
-    - --enable-admission-plugins=NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultTolerationSeconds,DefaultStorageClass,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota,Priority,PodSecurityPolicy
+    - --enable-admission-plugins=NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultTolerationSeconds,DefaultStorageClass,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota,Priority,PodSecurityPolicy,NodeRestriction
```
`NodeRestriction` admission plugin is not part of TLS bootstrapping. We should add it in separate commit or PR.

invidian commented on assets/lokomotive-kubernetes/bootkube/versions.tf:
```diff
@@ -6,5 +6,6 @@ terraform {
     local    = "~> 1.2"
     template = "~> 2.1"
     tls      = "~> 2.0"
+    random   = "~> 2.2"
```
I think this should be added together with it's being used. Otherwise it's a commit with no context.

surajssd replied:
It is used in packet, aws and bare-metal all of them.

invidian commented on assets/lokomotive-kubernetes/bootkube/variables.tf:
```diff
@@ -162,3 +162,8 @@ variable "blocked_metadata_cidrs" {
+
+variable "bootstrap_tokens" {
+  description = "List of boostrap tokens for all the nodes in the cluster in the form token-id and token-secret."
+  type        = list(any)
```
Why not `list(object(...))` ? Then mapping is not needed when passed to the template

invidian commented on assets/lokomotive-kubernetes/bootkube/assets.tf:
```diff
@@ -88,6 +88,28 @@ resource "local_file" "kubernetes" {
+resource "local_file" "bootstrap-secrets" {
+  filename = "${var.asset_dir}/charts/kube-system/bootstrap-secrets.yaml"
+  content = templatefile("${path.module}/resources/charts/bootstrap-secrets.yaml", {
+    bootstrap_tokens = [
```
With proper variable declaration, mapping here is not needed.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/bootstrap-manifests/bootstrap-controller-manager.yaml:
```diff
@@ -18,6 +18,8 @@ spec:
     - --cluster-signing-cert-file=/etc/kubernetes/secrets/ca.crt
     - --cluster-signing-key-file=/etc/kubernetes/secrets/ca.key
     - --configure-cloud-routes=false
+    - --controllers=*,tokencleaner
```
We don't specify token expiry time, so `tokencleaner` controller is not needed anyway?

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/bootstrap-secrets/templates/bootstrap-secret.yaml:
```diff
@@ -0,0 +1,15 @@
+  name: bootstrap-token-{{ $token.tokenID }}
+  namespace: kube-system
+stringData:
+  description: "Lokomotive generated bootstrap token"
```
We may consider adding expiry time here as well.

surajssd replied:
How do we renew that?

invidian replied:
With `terraform taint`? If it's out of scope, we should remove the cleaner controller and create an issue for it.

invidian commented on assets/lokomotive-kubernetes/bootkube/resources/charts/bootstrap-secrets/values.yaml:
```diff
@@ -0,0 +1,3 @@
+bootstrapTokens:
+- tokenID: ""
+  tokenSecret: ""
```
```suggestion
bootstrapTokens: []
#- tokenID: ""
#  tokenSecret: ""
```

This way, rendering the chart with no values provided will produce only valid objects. Also if we add validation rules, they should reject empty `tokenID` etc.

surajssd replied:
What do you mean by `validation rules` ?

invidian replied:
It's possible to implement values validation rules in chart itself, using the templates. Maybe at some point it would make sense to add such and then empty `tokenID` would be an invalid value.

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+}
+
+# Generate a cryptographically random token id (public)
+resource random_string "bootstrap-token-id" {
```
Maybe we should have a module generating the tokens, so we don't duplicate so much Terraform code?

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/outputs.tf:
```diff
@@ -27,6 +31,10 @@ output "calico_values" {
   value = module.bootkube.calico_values
 }
 
+output "bootstrap-secrets_values" {
```
As in #622:
```suggestion
output "bootstrap_secrets_values" {
```

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,40 @@
+}
+
+# Generate a cryptographically random token secret
+resource random_string "bootstrap-token-secret" {
```
Also as in #622:
```suggestion
resource random_string "bootstrap_token_secret" {
```

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/bootstrap-configs.tf:
```diff
@@ -0,0 +1,39 @@
+locals {
+  worker_bootstrap_tokens = [
```
We should have one token per worker pool, so all instances in autoscaling group use exactly same configuration.

invidian commented on assets/lokomotive-kubernetes/packet/flatcar-linux/kubernetes/workers/variables.tf:
```diff
@@ -94,8 +94,13 @@ variable "cluster_domain_suffix" {
 
-variable "kubeconfig" {
-  description = "Kubeconfig file"
+variable "ca_cert" {
```
Must this be base64 encoded? Or should we base64 encode it in the template?

surajssd replied:
It is already done here https://github.com/kinvolk/lokomotive/blob/e2499b44b06aadd76cd944250a9cdfa2d104ec93/assets/lokomotive-kubernetes/bootkube/outputs.tf#L49-L51

invidian replied:
Ah, thanks :+1:

invidian commented on cli/cmd/cluster-apply.go:
```diff
@@ -80,6 +80,18 @@ func runClusterApply(cmd *cobra.Command, args []string) {
 
 	kubeconfigPath := assetsKubeconfig(assetDir)
+
+	cu := controlplaneUpdater{
```
IMO this commit message is poorly written. This commit **adds update support for bootstrap-secrets controlplane component**, which implementation detail is that it should occur before checking if cluster nodes are available.

invidian commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -118,6 +118,12 @@ systemd:
         WantedBy=multi-user.target
 storage:
   files:
+    - path: /etc/kubernetes/kubeconfig
```
This commit implicitly changes how we deliver `kubeconfig` file to controller nodes on AWS without any explanation. This has been changed, as it was causing issues in the past. What is the reasoning behind this change? I think it should be done separately from TLS bootstrapping and well explained.

invidian commented on assets/lokomotive-kubernetes/bare-metal/flatcar-linux/kubernetes/bootkube.tf:
```diff
@@ -20,4 +20,6 @@ module "bootkube" {
   # Disable the self hosted kubelet.
   disable_self_hosted_kubelet = var.disable_self_hosted_kubelet
+
+  bootstrap_tokens = concat(local.controller_bootstrap_tokens, local.worker_bootstrap_tokens)
```
(commit message): the platform name is `bare-metal`, not `baremetal`.

invidian commented:
@surajssd what's the status of this PR?

surajssd commented:
> @surajssd what's the status of this PR?

started working on it again.

invidian commented:
Needs rebasing :/

surajssd commented:
Something is wrong with the AWS CI, this pipeline is failing consistently on the same test case. I wonder what changed that this test fails :thinking: :

```log
    ...
    TestNoExtraSSHKeysOnNodes: util.go:138: daemonset: test-ssh-keys-rfcmx, replicas: 4/4
    TestNoExtraSSHKeysOnNodes: util.go:138: daemonset: test-ssh-keys-rfcmx, replicas: 4/4
    TestNoExtraSSHKeysOnNodes: util.go:272: test-ssh-keys-rfcmx-2qvbk/test-ssh-keys: sha512sum: /home/core/.ssh/authorized_keys: No such file or directory
    TestNoExtraSSHKeysOnNodes: util.go:272: test-ssh-keys-rfcmx-ckhdl/test-ssh-keys: sha512sum: /home/core/.ssh/authorized_keys: No such file or directory
    TestNoExtraSSHKeysOnNodes: util.go:149: error while waiting for the daemonset: timed out waiting for the condition
```

surajssd commented:
This is from the pod running on the worker node, so it is true that there are no ssh keys there.

```console
$ kl test-ssh-keys-ttsht-qlkv6
uid=0(root) gid=0(root) groups=0(root)
total 16                                                                       
drwx------. 2  500  500 4096 Aug 19 13:55 .                                                                                                                   
drwxr-xr-x. 3 root root 4096 Aug 25 07:03 ..
sha512sum: /home/core/.ssh/authorized_keys: No such file or directory
```

invidian commented on assets/charts/control-plane/kubernetes/templates/kube-controller-manager.yaml:
```diff
@@ -60,6 +60,8 @@ spec:
         - --cluster-signing-key-file=/etc/kubernetes/secrets/ca.key
         - --configure-cloud-routes=false
+        - --controllers=*,tokencleaner
+        - --experimental-cluster-signing-duration=1h
```
I wouldn't enable token cleaning. If we clean them after an hour, it means that after 1h, worker pool will no longer be able to recover if instance gets recreated :/

invidian commented on assets/charts/control-plane/kubernetes/templates/bootstrap-cluster-role-binding.yaml:
```diff
@@ -0,0 +1,41 @@
+# enable bootstrapping nodes to create CSR
```
```suggestion
# Enable bootstrapping nodes to create CSR.
```

invidian commented on assets/charts/control-plane/kubernetes/templates/bootstrap-cluster-role-binding.yaml:
```diff
@@ -0,0 +1,41 @@
+  name: system:node-bootstrapper
+  apiGroup: rbac.authorization.k8s.io
+---
+# Approve all CSRs for the group "system:bootstrappers"
```
```suggestion
# Approve all CSRs for the group "system:bootstrappers".
```

invidian commented on assets/charts/control-plane/kubernetes/templates/bootstrap-cluster-role-binding.yaml:
```diff
@@ -0,0 +1,41 @@
+  name: system:certificates.k8s.io:certificatesigningrequests:nodeclient
+  apiGroup: rbac.authorization.k8s.io
+---
+# Approve renewal CSRs for the group "system:nodes"
```
```suggestion
# Approve renewal CSRs for the group "system:nodes".
```

invidian commented on assets/charts/control-plane/bootstrap-secrets/templates/bootstrap-secret.yaml:
```diff
@@ -0,0 +1,15 @@
+kind: Secret
+type: bootstrap.kubernetes.io/token
+metadata:
+  # Name MUST be of form "bootstrap-token-<token_id>"
```
```suggestion
  # Name MUST be of form "bootstrap-token-<token_id>".
```

invidian reviewed (changes requested):
LGTM, with some small nits. The one thing I would really re-consider is cleaning up the bootstrapping tokens.

invidian commented:
Also, rebase is needed again :/

invidian reviewed (approved):
LGTM, nice work @surajssd

surajssd commented:
I installed cluster with old version and tried to upgrade with this version after adding a variable `enable_tls_bootstrap = false` and it seems that it is trying to recreate controller nodes. I am investigating that now.

invidian commented:
>I installed cluster with old version and tried to upgrade with this version after adding a variable enable_tls_bootstrap = false and it seems that it is trying to recreate controller nodes. I am investigating that now.

@surajssd did you resolve this issue? If so, how? (GitHub does not show diff between the reviews :/)

surajssd commented:
@invidian yes resolved it by adding if else in helm chart config and in terraform config.

invidian reviewed (approved):
LGTM

knrt10 commented on assets/terraform-modules/packet/flatcar-linux/kubernetes/bootstrap-configs.tf:
```diff
@@ -0,0 +1,24 @@
+}
+
+# Generate a cryptographically random token id (public).
+resource random_string "bootstrap_token_id" {
+  count = var.enable_tls_bootstrap == true ? 1 : 0
+
+  length  = 6
+  upper   = false
+  special = false
+}
```
```suggestion
resource random_string "bootstrap_token_id" {
  count   = var.enable_tls_bootstrap == true ? 1 : 0
  length  = 6
  upper   = false
  special = false
}
```

Just a suggestion, totally up to you.

knrt10 replied:
Actually no need to. It would talk unnecessary extra work

knrt10 commented on assets/terraform-modules/bootkube/assets.tf:
```diff
@@ -79,6 +81,23 @@ resource "local_file" "kubernetes" {
+  })
+}
+
+# Populate kubernetes chart values file named kubernetes.yaml.
```
```suggestion
# Populate kubernetes chart values file named bootstrap-secrets.yaml.
```

surajssd replied:
good catch :+1:

knrt10 reviewed (commented):
Just NIT suggestion, rest looks great to me. Also, can you also please update the docs. Thanks

surajssd commented:
@knrt10 added docs.

knrt10 reviewed (approved):
Thanks for the effort. Looks great.

surajssd commented:
Thanks for the continuous feedback @invidian
